src/main/java/com/mongodb/hibernate/jdbc/MongoStatement.java

```diff
@@ -228,21 +279,379 @@ void checkClosed() throws SQLException {
+        }
+        return EMPTY_BATCH_RESULT;
+    }
+
+    private static boolean isTimeoutException(MongoException exception) {
+        // We do not check for `MongoExecutionTimeoutException` and `MongoOperationTimeoutException` here,
+        // because it is handled via error codes.
+        return exception instanceof MongoSocketReadTimeoutException
+                || exception instanceof MongoSocketWriteTimeoutException
+                || exception instanceof MongoTimeoutException;
+    }
+
+    enum CommandDescription {
+        INSERT("insert", false, true),
+        UPDATE("update", false, true),
+        DELETE("delete", false, true),
+        AGGREGATE("aggregate", true, false);
+
+        private final String commandName;
+        private final boolean returnsResultSet;
+        private final boolean isUpdate;
+
+        CommandDescription(String commandName, boolean returnsResultSet, boolean isUpdate) {
+            this.commandName = commandName;
+            this.returnsResultSet = returnsResultSet;
+            this.isUpdate = isUpdate;
+        }
+
+        String getCommandName() {
+            return commandName;
+        }
+
+        /**
+         * Indicates whether the command is used in {@code executeUpdate(...)} or {@code executeBatch()} methods.
+         *
+         * @return true if the command is used in update operations, false if it is used in query operations.
+         */
+        boolean isUpdate() {
+            return isUpdate;
+        }
+
+        /**
+         * Indicates whether the command returns a {@link ResultSet}.
+         *
+         * @see #executeQuery(String)
+         */
+        boolean returnsResultSet() {
+            return returnsResultSet;
+        }
+
+        static CommandDescription fromString(String commandName) throws SQLFeatureNotSupportedException {
+            return switch (commandName) {
+                case "insert" -> INSERT;
+                case "update" -> UPDATE;
+                case "delete" -> DELETE;
+                case "aggregate" -> AGGREGATE;
+                default -> throw new SQLFeatureNotSupportedException("Unsupported command: %s".formatted(commandName));
+            };
+        }
+    }
+
+    private static class WriteModelConverter {
+        private static final String UNSUPPORTED_MESSAGE_STATEMENT_FIELD = "Unsupported field in %s statement: [%s]";
+        private static final String UNSUPPORTED_MESSAGE_COMMAND_FIELD = "Unsupported field in %s command: [%s]";
+
+        private static final Set<String> SUPPORTED_INSERT_COMMAND_FIELDS = Set.of("documents");
+
+        private static final Set<String> SUPPORTED_UPDATE_COMMAND_FIELDS = Set.of("updates");
+        private static final Set<String> SUPPORTED_UPDATE_STATEMENT_FIELDS = Set.of("q", "u", "multi");
+
+        private static final Set<String> SUPPORTED_DELETE_COMMAND_FIELDS = Set.of("deletes");
+        private static final Set<String> SUPPORTED_DELETE_STATEMENT_FIELDS = Set.of("q", "limit");
```

Comment:
[optional]

What do you think about moving these to `CommandDescription`? That way the information tightly related to commands stays mostly there.

P.S. At first I thought about moving them there as instance methods, potentially with the checks, but then I realized that different commands have different structure (for example, `aggregate` does not have statements), which makes such endeavour non-trivial.